module/zfs/zpl_ctldir.c

```diff
@@ -99,18 +99,35 @@ zpl_root_readdir(struct file *filp, void *dirent, filldir_t filldir)
  * Get root directory attributes.
  */
 /* ARGSUSED */
+
 static int
-zpl_root_getattr(struct vfsmount *mnt, struct dentry *dentry,
-    struct kstat *stat)
+zpl_root_getattr(
+#if defined(HAVE_3ARGS_IOPS_GETATTR)
+    struct vfsmount *mnt,
+    struct dentry *dentry,
+    struct kstat *stat
+#elif defined(HAVE_5ARGS_IOPS_GETATTR)
+    const struct path *path,
+    struct dentry *dentry,
+    struct kstat *stat,
+    u32 request_mask,
+    unsigned int query_flags
+#endif /* defined(HAVE_PATH_SIMPLE_GETATTR) */
+)
```

Comment:
@tuxoko,
Thanks for the feedback and pointing out the dentry error.

I'm not certain I understand what you mean.  Using this function as an example, it must provide the same interface as simple_getattr() and also calls simple_getattr() within its body, so it seems to me that one either has conditional blocks like this, or separate implementations so that the arguments received can be passed on to the simple_getattr() call.

That said, I agree this is not easy to read and likely not the best way.  Brian pointed out that in some cases I could move the common logic into an _impl function and then have one version with the old interface and one with the new, wrapped in the appropriate ifdefs, that both call the same common function after making the simple_getattr() call appropriately.

I just pushed an update that fixes the mistakes that earlier were preventing it from building.  I'll work on it more tomorrow.